### DISCUSSION TITLE
Adapter Configuration - Expert of Mixture

### DISCUSSION THREAD

**Discussion by simon-lund** [adapter-hub/adapters#627] [Ideas]
Hello,

I am a student of computer science at the LMU Munich and I am working on fine-tuning vision-language models in a multitask setting as part of my master thesis. The focus of my work is on Mixture of Expert models, which select and combine relevant adapters for different inputs.

During my research, I came across your adapter framework. I have already read some of the documentation, but I am still unsure whether it would be possible to “simply” add such a meta module. I would like to contribute to this framework with such a module, but I would appreciate a little help on where to start.

Beyond that, I would like to ask whether existing models can be extended. Specifically, a combination of CLIP + Llama would allow me to rebuild the code of a paper ([Octavius](https://arxiv.org/pdf/2311.02684.pdf)).

**Comment by calpt** [comment]
Hey @simon-lund!

Thanks for reaching out, this sounds very interesting. We'd be happy to see your work integrated into our library. Haven't studied your topic and the linked paper in detail yet, so these are only preliminary high-level answers:

> During my research, I came across your adapter framework. I have already read some of the documentation, but I am still unsure whether it would be possible to “simply” add such a meta module. I would like to contribute to this framework with such a module, but I would appreciate a little help on where to start.

Yes, in principle it should be possible to integrate such new modules into the library. I think what you're trying to achieve most closely resembles a new [composition block](https://docs.adapterhub.ml/adapter_composition.html#composition-blocks-overview) and could be implemented as a new composition block type, so I'll try to give some context here:
- all composition block implementations are based on a common class mixin [here](https://github.com/adapter-hub/adapters/blob/b972812009e937d5bee5fb04cb6b29486171beb5/src/adapters/methods/adapter_layer_base.py#L148).
- each composition block comes with a method defined in this class, which is called during the model forward pass by each layer containing adapter modules. Your use case might be somewhat related to Average composition, implemented [here](https://github.com/adapter-hub/adapters/blob/b972812009e937d5bee5fb04cb6b29486171beb5/src/adapters/methods/adapter_layer_base.py#L434C9-L434C9). Maybe it makes sense to use this composition block as orientation in the beginning.
- the abstract composition class mixin is implemented by method-specific mixins for each adapter method, e.g. for LoRA [here](https://github.com/adapter-hub/adapters/blob/b972812009e937d5bee5fb04cb6b29486171beb5/src/adapters/methods/lora.py#L286). These specific mixins might customize the general implementation to each specific adapter method.
- composition blocks are automatically integrated into all currently supported models, so ideally it should not be necessary to modify model implementations directly when adding a new composition block

> Beyond that, I would like to ask whether existing models can be extended. Specifically, a combination of CLIP + Llama would allow me to rebuild the code of a paper ([Octavius](https://arxiv.org/pdf/2311.02684.pdf)).

In general, both CLIP and Llama are already [supported by our library](https://docs.adapterhub.ml/model_overview.html). In principle it should be possible to compose these models using standard methods provided in Transformers, e.g. joining vision encoders and text decoders as described [here](https://huggingface.co/docs/transformers/model_doc/vision-encoder-decoder).
This might require some additional tweaking to work well with adapters though.

In general, our contributing guides on [adding new model support](https://docs.adapterhub.ml/contributing/adding_adapters_to_a_model.html) and [adding new adapter methods](https://docs.adapterhub.ml/contributing/adding_adapter_methods.html) might also provide helpful context to this regard.

Hope this is somewhat helpful as general pointers. Happy to help with more specific questions/ issues related to your concrete use case!


---

edit: moved this to our discussions forum

**Reply by simon-lund** [reply]
Hello, 

thank you very much for your explanation, and I want to apologize for the late response.
I like this library very much and find it very useful, however, I decided to work with the PEFT library for now, so I don't have to adapt the model but can focus on my work. Once I am done with this, I will come back to this discussion since I am still interested in contributing to this library.

Best
Simon
